### PR TITLE
[ONNX] Add `acceptable_error_percentage` to backend tests

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82623
* __->__ #82622
* #82621

To enable more test coverage for numerically unstable tests that
occasionally fails with sporadic element mismatch. Currently these
tests have to be disabled in CI. This PR provides a configurable
tolerance threshold for such issues.

### CODE REVIEW

File: torch/onnx/verification.py [hunk 1]
```diff
@@ -113,14 +113,29 @@ def _ort_session(
...
-def _compare_ort_pytorch_outputs(ort_outs, pt_outs, rtol, atol):
+def _compare_ort_pytorch_outputs(
+    ort_outs, pt_outs, rtol, atol, acceptable_error_percentage: Optional[float]
+):
     pt_outs, _ = torch.jit._flatten(pt_outs)
     pt_outs = _unpack_to_numpy(pt_outs)
 
     assert len(pt_outs) == len(ort_outs), "number of outputs differ"
+    if acceptable_error_percentage:
+        assert (
```
Comment (justinchuby) [inline torch/onnx/verification.py:124]: I would raise a ValueError for user mistakes.

File: test/onnx/test_models.py [hunk 2]
```diff
@@ -143,13 +143,11 @@ def test_resnet(self):
         x = Variable(torch.randn(BATCH_SIZE, 3, 224, 224).fill_(1.0))
         self.exportTest(toC(resnet50()), toC(x), atol=1e-6)
 
-    @unittest.skip(
-        "This test has been flaky on trunk and PRs. See https://github.com/pytorch/pytorch/issues/79540"
-    )
     @skipScriptTest(min_opset_version=15)  # None type in outputs
+    # This test is numerically unstable. Sporadic single element mismatch occurs occasionally.
```
Comment (justinchuby) [inline test/onnx/test_models.py:147]: Trying to understand context: Is single element mismatch not a bug?

Comment (BowenBao) [reply test/onnx/test_models.py:147]: Good question. I think the backend folks would have a better say. It's definitely in a gray area. But I wouldn't want to disable the entire test until everything is fixed either. 
rtol/atol won't work in this case. It's a kind of "spike" mismatch, very large diff in value.

File: torch/onnx/verification.py [hunk 3]
```diff
@@ -113,14 +113,29 @@ def _ort_session(
     return ort_session
 
 
-def _compare_ort_pytorch_outputs(ort_outs, pt_outs, rtol, atol):
+def _compare_ort_pytorch_outputs(
+    ort_outs, pt_outs, rtol, atol, acceptable_error_percentage: Optional[float]
+):
```
Comment (justinchuby) [inline torch/onnx/verification.py:118]: It'd be helpful to define and explain `acceptable_error_percentage` in the docstring here.

File: test/onnx/test_models.py (hunk 2 above)
Comment (justinchuby) [reply test/onnx/test_models.py:147]: Gotcha!

**Review by justinchuby — APPROVED** [review]

File: torch/onnx/verification.py (hunk 3 above)
Comment (BowenBao) [reply torch/onnx/verification.py:118]: done

File: torch/onnx/verification.py [hunk 4]
```diff
@@ -116,14 +116,34 @@ def _ort_session(
...
 
     assert len(pt_outs) == len(ort_outs), "number of outputs differ"
-    if acceptable_error_percentage:
-        assert (
-            acceptable_error_percentage <= 1.0 and acceptable_error_percentage >= 0.0
-        ), "if set, acceptable_error_percentage should be between 0.0 and 1.0"
+    if (
+        acceptable_error_percentage
+        and acceptable_error_percentage <= 1.0
+        and acceptable_error_percentage >= 0.0
```
Comment (justinchuby) [inline torch/onnx/verification.py:142]: Should the logic be reversed?

**Comment by BowenBao** [comment]
@abock for merge approval

File: torch/onnx/verification.py [hunk 5]
```diff
@@ -147,7 +155,18 @@ def _compare_ort_pytorch_outputs(
 
     for ort_out, pt_out in zip(ort_outs, pt_outs):
         try:
-            np.testing.assert_allclose(ort_out, pt_out, rtol=rtol, atol=atol)
+            # TODO: Remove `check_shape` option once every shape inconsistent issue is addressed.
+            if not check_shape:
```
Comment (justinchuby) [inline torch/onnx/verification.py:159]: Indent?

File: torch/onnx/verification.py [hunk 6]
```diff
@@ -122,27 +122,59 @@ def _compare_ort_pytorch_outputs(
...
+                check_dtype=check_dtype,
+                equal_nan=True,
+            )
+        except AssertionError as e:
+            if acceptable_error_percentage:
+                error_percentage = 1 - np.sum(
+                    np.isclose(ort_out, pt_out, rtol=rtol, atol=atol)
+                ) / np.prod(ort_out.shape)
+                if error_percentage <= acceptable_error_percentage:
+                    continue
```
Comment (justinchuby) [inline torch/onnx/verification.py:176]: I think it would be helpful to produce a warning here

File: torch/onnx/verification.py [hunk 7]
```diff
@@ -122,27 +122,59 @@ def _compare_ort_pytorch_outputs(
...
+            raise e
```
Comment (justinchuby) [inline torch/onnx/verification.py:177]:
```suggestion
            raise
```

File: torch/onnx/verification.py (hunk 6 above)
Comment (abock) [reply torch/onnx/verification.py:176]: We should definitely log a warning and probably coalesce similar warnings into a summary warning if log pollution is an issue.

**Review by abock — CHANGES_REQUESTED** [review]
@justinchuby has good feedback here

**Review by abock — CHANGES_REQUESTED** [review]
@justinchuby has good feedback here

File: torch/onnx/verification.py (hunk 6 above)
Comment (BowenBao) [reply torch/onnx/verification.py:176]: Agreed. I will be adding a warning for now.

**Review by abock — APPROVED** [review]

**Pull request closed without merging**